### PR TITLE
refactor datum's as_bool and as_int

### PR DESCRIPTION
**Pull request by H-ZeX**
Signed-off-by: H-ZeX <hzx20112012@gmail.com>

## What have you changed? (mandatory)

Refactor `Datum::as_bool` and `Datum::as_int` base on convert_*_to_* functions and TiDB's implement.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Unit test.

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

### CODE REVIEW

File: src/coprocessor/dag/expr/column.rs [hunk 1]
```diff
@@ -17,8 +17,8 @@ impl Column {
     }
 
     #[inline]
-    pub fn eval_int(&self, row: &[Datum]) -> Result<Option<i64>> {
-        row[self.offset].as_int()
+    pub fn eval_int(&self, row: &[Datum], ctx: &mut EvalContext) -> Result<Option<i64>> {
```
Comment (breeswish) [inline src/coprocessor/dag/expr/column.rs]: ctx should be placed before the `row` parameter.

**Comment by breeswish** [comment]
Unit tests failed:

```
failures:

    coprocessor::dag::expr::builtin_arithmetic::tests::test_arithmetic_int

    coprocessor::dag::expr::builtin_arithmetic::tests::test_multiply_int_unsigned

    coprocessor::dag::expr::builtin_math::tests::test_abs

    coprocessor::dag::expr::builtin_math::tests::test_ceil

    coprocessor::dag::expr::builtin_math::tests::test_floor

    coprocessor::dag::expr::builtin_math::tests::test_truncate_decimal

    coprocessor::dag::expr::builtin_math::tests::test_truncate_int

    coprocessor::dag::expr::builtin_math::tests::test_truncate_real

    coprocessor::dag::expr::builtin_op::tests::test_unary_minus_op

    coprocessor::dag::expr::builtin_other::tests::test_bit_count

    coprocessor::dag::expr::builtin_string::tests::test_left

    coprocessor::dag::expr::builtin_string::tests::test_left_binary

    coprocessor::dag::expr::builtin_string::tests::test_right

    coprocessor::dag::expr::builtin_string::tests::test_right_binary

    coprocessor::dag::expr::builtin_string::tests::test_space

    coprocessor::dag::expr::builtin_string::tests::test_substring_2_args

    coprocessor::dag::expr::builtin_string::tests::test_substring_3_args

    coprocessor::dag::expr::builtin_string::tests::test_substring_binary_2_args

    coprocessor::dag::expr::builtin_string::tests::test_substring_binary_3_args

    coprocessor::dag::expr::builtin_string::tests::test_substring_index

    coprocessor::dag::expr::column::tests::test_column_eval

    coprocessor::dag::expr::constant::tests::test_constant_eval

    coprocessor::dag::expr::tests::test_expression_eval

```

**Comment by breeswish** [comment]
CI failed again :(

**Comment by breeswish** [comment]
Mostly fine. But I can't understand the purpose of refactoring datum's interface to use our cast infra (talking about `as_int`). Currently all push down functions except for aggregate functions use explicit casts, so looks like there is no need to implement implicit cast in the datum? @lonng PTAL

**Comment by lonng** [comment]
@breeswish We should complement the `Datum::as_int` to cover all enum variants if we want to push-down the CAST functions (only cover `Datum::U64` and `Datum::I64` in the current implementation, see: https://github.com/tikv/tikv/pull/5113/files#diff-96a987fe4cdf69f2443ef606552429a9R42).

And we refactor all CAST related functions not just for the aggregate functions but also for all CAST scenario in the future (e.g: `select * from t1 where t1.a == 10` will contain a cast if the `t1.a` is not an integer).

File: src/coprocessor/dag/expr/constant.rs [hunk 2]
```diff
@@ -3,17 +3,72 @@
...
 use crate::coprocessor::codec::mysql::{Decimal, Duration, Json, Time};
-use crate::coprocessor::codec::Datum;
+use crate::coprocessor::codec::{Datum, Error};
+use crate::coprocessor::dag::expr::EvalContext;
+use cop_datatype::FieldTypeTp;
 
 impl Datum {
     #[inline]
-    pub fn as_int(&self) -> Result<Option<i64>> {
+    pub fn as_bool(&self, ctx: &mut EvalContext) -> Result<Option<bool>> {
```
Comment (lonng) [inline src/coprocessor/dag/expr/constant.rs]: I think we should implement `AsMySQLBool` trait for `Datum` instead of adding an extra method to it.

Comment (H-ZeX) [reply src/coprocessor/dag/expr/constant.rs]: The AsMySQLBool funct is `fn as_mysql_bool(&self, context: &mut EvalContext) -> Result<bool>;`, its `Result` is `tikv::coprocessor::error`. The Datum's `as_*`'s `Result<Option<x>>`'s Err is `tikv::coprocessor::codec::error`

Comment (lonng) [reply src/coprocessor/dag/expr/constant.rs]: The `tikv::coprocessor::codec::error` will convert to `tikv::coprocessor::error` eventually.

**Comment by breeswish** [comment]
@lonng 

"We should complement the Datum::as_int to cover all enum variants if we want to push-down the CAST functions"

Why not directly implement call cast functions in such secnario? For example, for int -> real cast, maybe:

```rust
let val = xxx.eval_int(); // no cast
let casted = cast_int_as_real(val); // explicit cast
return casted
```

It looks clearer than implicit cast like `return xxx.eval_real()` where `xxx` is in fact received as a `int`.

**Comment by lonng** [comment]
@breeswish Get your point. @H-ZeX I think you could take him suggestion into account.

**Pull request closed without merging**